synapse/config/oembed.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import logging
+import re
+from os import listdir, path
+from urllib.parse import urlparse, urlunparse
+
+from ._base import Config
+from ._util import validate_config
+
+logger = logging.Logger(__name__)
+
+
+class OembedConfig(Config):
+    section = "oembed"
+    oembed_providers = {}
+
+    def read_config(self, config, **kwargs):
+        oembed_dir = config.get("oembed_providers_dir")
+
+        if not oembed_dir:
+            return
+
+        oembed_providers = []
+        try:
+            for fname in listdir(oembed_dir):
+                if fname.endswith(".json"):
+                    fpath = path.join(oembed_dir, fname)
+                    try:
+                        with open(fpath) as f:
+                            oembed_providers += json.loads(f.read())
+                    except Exception:
+                        logger.exception(fpath)
+        except Exception:
+            logger.exception(oembed_dir)
+
+        if not oembed_providers:
+            return
+
+        try:
+            validate_config(
+                _OEMBED_SCHEMA,
+                oembed_providers,
+                ("oembed_providers",),
+            )
+        except Exception:
+            pass
+            # logger.exception('oembed_providers')
+            # and go on, this is not a show stopper
+            # return
+
+        for provider in oembed_providers:
+            provider_url = provider["provider_url"].rstrip("/")
+            for endpoint in provider["endpoints"]:
+                if "schemes" not in endpoint:
+                    continue
+
+                patterns = []
+                for s in endpoint["schemes"]:
+                    results = urlparse(s)
+                    pattern = urlunparse(
+                        [
+                            results.scheme,
+                            re.escape(results.netloc).replace("\\*", "[a-zA-Z0-9_-]+"),
+                        ]
+                        + [re.escape(part).replace("\\*", ".+") for part in results[2:]]
+                    )
+                    patterns.append(re.compile(pattern))
+                endpoint["patterns"] = patterns
```

Comment:
it seems poor style to modify the original `oembed_providers` data structure, as well as being wasteful (we will keep a lot of information in memory that will never be used). I suggest building a completely new structure.

Reply:
I have replaced oembed_providers with oembed_endpoints, should be a bit leaner.